src/lib/modules/blockchain_process/proxy.js

```diff
@@ -1,156 +1,178 @@
-
-  proxy.on('error', function (e) {
-    console.error(__("Error forwarding requests to blockchain/simulator"), e.message);
-  });
 
-  proxy.on('proxyRes', (proxyRes) => {
-    let resBody = [];
-    proxyRes.on('data', (b) => resBody.push(b));
-    proxyRes.on('end', function () {
-      resBody = Buffer.concat(resBody).toString();
-      if (resBody) {
-        parseResponse(ipc, resBody);
+    onProxyReq(proxyReq, req, _res) {
+      if (req.method === 'POST') {
+        // messages TO the target
+        Asm.connectTo(chain([
+          req,
+          parser()
+        ])).on('done', asm => {
+          trackRequest(asm.current);
+        });
       }
-    });
-  });
-
-  let server = http.createServer((req, res) => {
-    let reqBody = [];
-    req.on('data', (b) => {
-      reqBody.push(b);
-    })
-      .on('end', () => {
-        reqBody = Buffer.concat(reqBody).toString();
-        if (reqBody) {
-          parseRequest(reqBody);
-        }
-      });
+    },
 
-    if (!ws) {
-      proxy.web(req, res);
+    onProxyRes(proxyRes, req, _res) {
+      if (req.method === 'POST') {
+        // messages FROM the target
+        Asm.connectTo(chain([
+          proxyRes,
+          parser()
+        ])).on('done', asm => {
+          trackResponse(asm.current);
+        });
+      }
     }
-  });
+  };
 
   if (ws) {
-    const WsParser = require('simples/lib/parsers/ws'); // npm install simples
-
-    server.on('upgrade', function (req, socket, head) {
-      proxy.ws(req, socket, head);
-    });
-
-    proxy.on('open', (proxySocket) => {
-      proxySocket.on('data', (data) => {
-        parseResponse(ipc, data.toString().substr(data.indexOf("{")));
+    proxyOpts.onProxyReqWs = (_proxyReq, _req, socket, _options, _head) => {
+      // messages TO the target
+      const wsp = new WsParser(0, false);
+      wsp.on('frame', (frame) => {
+        trackRequest(parseJsonMaybe(frame.data.toString()));
       });
-    });
+      pump(cloneable(socket), wsp);
+    };
 
-    proxy.on('proxyReqWs', (proxyReq, req, socket) => {
-      var parser = new WsParser(0, false);
-      socket.pipe(parser);
-      parser.on('frame', function (frame) {
-        parseRequest(frame.data);
+    proxyOpts.onOpen = (proxySocket) => {
+      // messages FROM the target
+      const recv = new WebSocket.Receiver();
+      recv.on('message', (data) => {
+        trackResponse(parseJsonMaybe(data));
       });
-
-    });
+      pump(cloneable(proxySocket), recv);
+    };
   }
-  const listenPort = port - constants.blockchain.servicePortOnProxy;
+
+  const proxy = proxyMiddleware(proxyOpts);
+  const app = express();
+  app.use('*', proxy);
+
   return new Promise(resolve => {
-    server.listen(listenPort, defaultHost, () => {
-      resolve(server);
-    });
+    const server = app.listen(
+      port - constants.blockchain.servicePortOnProxy,
+      defaultHost,
+      () => { resolve(server); }
+    );
+    if (ws) {
+      server.on('upgrade', proxy.upgrade);
```

Comment:
Out of curiosity, what is this one doing?

Reply:
The `http-proxy-middleware` docs describe it in the section *[External WebSocket upgrade](https://github.com/chimurai/http-proxy-middleware#external-websocket-upgrade)*. Without the external upgrade, the connections weren't always working, I'm not sure why. However, I've made a number of changes since I first added that code and I'm going to try again without it.

Reply:
It seems that if the external upgrade is not performed then in the dashboard embark never progresses past `Blockchain node is ready` to `Ethereum node detected..`.

Reply:
This web3 request in `blockchain_connector/provider.js` (possibly others later on) requires the external upgrade or else it eventually times out. In the meantime, other requests are made successfully.

https://github.com/embark-framework/embark/blob/master/src/lib/modules/blockchain_connector/provider.js#L56

I used the debugger to take a look at the headers to see if I could spot the problem, but I haven't figured it out. The external upgrade works, so I think we can just rely on it for now. Thoughts?